File: LeetCode_ProblemSet/Go/Problem1905.go
package main 

func countSubIslands(grid1, grid2 [][]int) int {
    islands := getIslands(grid2)
	return getSubIslandCount(grid1, islands)
}

func getSubIslandCount(grid [][]int, islands [][][]int) int {
	count := 0
	for _, island := range islands {
		isSubIsland := true
		for _, cell := range island {
			if grid[cell[0]][cell[1]] == 0 {
				isSubIsland = false
				break
			}
		}
		if isSubIsland {
			count++
		}
	}
	return count
}

func getIslands(grid [][]int) [][][]int {
	m, n := len(grid), len(grid[0])
	islands := [][][]int{}
	q := [][]int{}

	for row := 0; row < m; row++ {
		for col := 0; col < n; col++ {
			if grid[row][col] == 0 {
				continue
			}
			island := [][]int{[]int{row, col}}
			grid[row][col] = 0
			q = append(q, []int{row, col})

			for len(q) > 0 {
				cell := q[0]
				q = q[1:]
				for _, neighbor := range getNeighbors(cell[0], cell[1], m, n) {
					n_r, n_c := neighbor[0], neighbor[1]
					if grid[n_r][n_c] == 0 {
						continue
					}
					grid[n_r][n_c] = 0
					q = append(q, []int{n_r, n_c})
					island = append(island, []int{n_r, n_c})
				}
			}
			islands = append(islands, island)
		}
	}
	return islands
}

func getNeighbors(r, c, m, n int) [][]int {
	changeX := []int{-1, 0, 0, 1}
	changeY := []int{0, -1, 1, 0}
	neighbors := [][]int{}
	for i := 0; i < 4; i++ {
		if 0 <= r + changeX[i] && r + changeX[i] < m &&
			0 <= c + changeY[i] && c + changeY[i] < n {
			neighbors = append(neighbors, []int{r + changeX[i], c + changeY[i]})
		}
	}
	return neighbors
}

File: LeetCode_ProblemSet/Problem1905.py
from collections import deque

class Solution(object):
    def countSubIslands(self, grid1, grid2):
        """
        :type grid1: List[List[int]]
        :type grid2: List[List[int]]
        :rtype: int
        """
        islands = self.getIslands(grid2)
        return self.getSubIslandCount(grid1, islands)
    
    def getSubIslandCount(self, grid, islands):
        count = 0
        for island in islands:
            isSubIsland = True
            for x, y in island:
                if grid[x][y] == 0:
                    isSubIsland = False
                    break
            if isSubIsland:
                count += 1
        return count
        
    def getIslands(self, grid):
        m, n = len(grid), len(grid[0])
        islands = []
        q = deque()
        
        for row in range(m):
            for col in range(n):
                if grid[row][col] == 0:
                    continue
                    
                island = [(row, col)]
                grid[row][col] = 0
                q.append((row, col))
                
                while len(q) > 0:
                    r, c = q.popleft()
                    for n_r, n_c in self.getNeighbors(r, c, m, n):
                        if grid[n_r][n_c] == 0:
                            continue
                        grid[n_r][n_c] = 0
                        island.append((n_r, n_c))
                        q.append((n_r, n_c))
                            
                islands.append(island)
                                
        return islands
                        
                        
    def getNeighbors(self, r, c, m, n):
        change_x = [-1, 0, 0, 1]
        change_y = [0, -1, 1, 0]
        neighbors = []
        
        for x, y in zip(change_x, change_y):
            if 0 <= r + x < m and 0 <= c + y < n:
                neighbors.append((r + x, c + y))
                
        return neighbors